File: src/com/lushprojects/circuitjs1/client/RoutedWireElm.java
/*
    Copyright (C) Paul Falstad and Iain Sharp

    This file is part of CircuitJS1.

    CircuitJS1 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    CircuitJS1 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CircuitJS1.  If not, see <http://www.gnu.org/licenses/>.
*/

package com.lushprojects.circuitjs1.client;

import java.util.ArrayList;

import com.google.gwt.xml.client.Document;
import com.google.gwt.xml.client.Element;

    class RoutedWireElm extends WireElm {
	ArrayList<Point> routePoints;

	public RoutedWireElm(int xx, int yy) { super(xx, yy); }
	public RoutedWireElm(int xa, int ya, int xb, int yb, int f,
			     StringTokenizer st) {
	    super(xa, ya, xb, yb, f, st);
	}

	public RoutedWireElm(ArrayList<Point> points) {
	    super(points.get(0).x, points.get(0).y);
	    setPoints(points);
	}

	int getDumpType() { return 0; }
	String getXmlDumpType() { return "rw"; }

	void dumpXml(Document doc, Element elem) {
	    super.dumpXml(doc, elem);
	    if (routePoints != null && routePoints.size() > 0) {
		StringBuilder sb = new StringBuilder();
		for (int i = 0; i < routePoints.size(); i++) {
		    if (i > 0)
			sb.append(';');
		    Point p = routePoints.get(i);
		    sb.append(p.x).append(',').append(p.y);
		}
		elem.appendChild(doc.createTextNode(sb.toString()));
	    }
	}

	void undumpXml(XMLDeserializer xml) {
	    super.undumpXml(xml);
	    try {
		String contents = xml.parseContents();
		if (contents != null && contents.length() > 0) {
		    ArrayList<Point> points = new ArrayList<Point>();
		    String[] pairs = contents.split(";");
		    for (String pair : pairs) {
			String[] xy = pair.split(",");
			points.add(new Point(Integer.parseInt(xy[0]), Integer.parseInt(xy[1])));
		    }
		    if (points.size() >= 2)
			routePoints = points;
		}
	    } catch (Exception e) {}
	}

	int getShortcut() { return 'W'; }

	void setPoints() {
	    setPoints(true);
	}

	void setPoints(boolean routing) {
	    super.setPoints();

	    // if endpoints haven't changed, keep existing route
	    if (routePoints != null && routePoints.size() >= 2) {
		Point first = routePoints.get(0);
		Point last = routePoints.get(routePoints.size() - 1);
		if (first.x == x && first.y == y && last.x == x2 && last.y == y2)
		    return;
	    }

	    if (!routing) {
		// just set up a straight line, no routing
		routePoints = new ArrayList<Point>();
		routePoints.add(point1);
		routePoints.add(point2);
		return;
	    }

	    WireRouter router = new WireRouter();
	    router.initGrid(this);

	    routePoints = router.routeWire(x, y, x2, y2);
	    if (routePoints.size() < 2) {
		sim.console("routing failed");
		// fallback: simple L-shape
		routePoints = new ArrayList<Point>();
		routePoints.add(point1);
		routePoints.add(new Point(x2, y));
		routePoints.add(point2);
		return;
	    }
	    sim.console("route success");
	}

	void setPoints(ArrayList<Point> points) {
	    Point first = points.get(0);
	    Point last = points.get(points.size() - 1);
	    x = first.x; y = first.y;
	    x2 = last.x; y2 = last.y;
	    setPoints(false);
	    routePoints = points;
	}

	// split this routed wire at the point nearest (mx, my).
	// returns the new second-half wire, or null if split is not possible.
	RoutedWireElm split(int mx, int my) {
	    if (routePoints == null || routePoints.size() < 2)
		return null;

	    // find the nearest segment
	    int bestSeg = -1;
	    int bestDist = Integer.MAX_VALUE;
	    for (int i = 0; i < routePoints.size() - 1; i++) {
		Point a = routePoints.get(i);
		Point b = routePoints.get(i + 1);
		int d = lineDistanceSq(a.x, a.y, b.x, b.y, mx, my);
		if (d < bestDist) {
		    bestDist = d;
		    bestSeg = i;
		}
	    }

	    Point a = routePoints.get(bestSeg);
	    Point b = routePoints.get(bestSeg + 1);

	    // snap the click point onto the segment (horizontal or vertical)
	    int sx, sy;
	    if (a.x == b.x) {
		sx = a.x;
		sy = snapGrid(my);
		sy = Math.max(Math.min(a.y, b.y), Math.min(sy, Math.max(a.y, b.y)));
	    } else {
		sy = a.y;
		sx = snapGrid(mx);
		sx = Math.max(Math.min(a.x, b.x), Math.min(sx, Math.max(a.x, b.x)));
	    }

	    // don't split at an existing endpoint
	    if ((sx == x && sy == y) || (sx == x2 && sy == y2))
		return null;

	    // build route points for the two halves
	    ArrayList<Point> rp1 = new ArrayList<Point>();
	    for (int i = 0; i <= bestSeg; i++)
		rp1.add(routePoints.get(i));
	    rp1.add(new Point(sx, sy));

	    ArrayList<Point> rp2 = new ArrayList<Point>();
	    rp2.add(new Point(sx, sy));
	    for (int i = bestSeg + 1; i < routePoints.size(); i++)
		rp2.add(routePoints.get(i));

	    // update this wire to be the first half
	    setPoints(rp1);

	    // create and return the second half
	    return new RoutedWireElm(rp2);
	}

	// reroute this wire through an intermediate point
	void rerouteVia(int vx, int vy) {
	    WireRouter router = new WireRouter();
	    router.initGrid(this);
	    ArrayList<Point> rp1 = router.routeWire(x, y, vx, vy);
	    if (rp1.size() < 2)
		return;

	    // write the first route into the grid so the second avoids it
	    for (int i = 0; i < rp1.size() - 1; i++) {
		Point a = rp1.get(i);
		Point b = rp1.get(i + 1);
		router.addWire(a.x, a.y, b.x, b.y);
	    }

	    ArrayList<Point> rp2 = router.routeWire(vx, vy, x2, y2);
	    if (rp2.size() < 2)
		return;

	    // combine the two routes, removing the duplicate via point
	    ArrayList<Point> combined = new ArrayList<Point>(rp1);
	    for (int i = 1; i < rp2.size(); i++)
		combined.add(rp2.get(i));
	    routePoints = combined;
	}

	void addRoutingObstacle(WireRouter router) {
	    if (routePoints == null)
		return;
	    for (int i = 0; i < routePoints.size() - 1; i++) {
		Point a = routePoints.get(i);
		Point b = routePoints.get(i + 1);
		router.addWire(a.x, a.y, b.x, b.y);
	    }
	}

	void draw(Graphics g) {
	    setVoltageColor(g, volts[0]);
	    if (busWidth > 1)
		g.setLineWidth(5.0);
	    for (int i = 0; i < routePoints.size() - 1; i++)
		drawThickLine(g, routePoints.get(i), routePoints.get(i + 1));
	    if (busWidth > 1)
		g.setLineWidth(1.0);
	    doDots(g);

	    final int m = 5;
	    int minX = point1.x, minY = point1.y, maxX = point1.x, maxY = point1.y;
	    for (int i = 1; i < routePoints.size(); i++) {
		Point p = routePoints.get(i);
		minX = min(minX, p.x);
		minY = min(minY, p.y);
		maxX = max(maxX, p.x);
		maxY = max(maxY, p.y);
	    }
	    setBbox(minX - m, minY - m, maxX + m, maxY + m);
	    String s = "";
	    if (busWidth > 1 && mustShowBusValue()) {
		int value = getBusValue();
		s = ""+value;
	    } else if (busWidth == 1) {
		if (mustShowCurrent())
		    s = getShortUnitText(Math.abs(getCurrent()), "A");
		if (mustShowVoltage())
		    s = (s.length() > 0 ? s + " " : "") + getShortUnitText(volts[0], "V");
	    }
	    if (s.length() > 0)
		drawValuesOnLongestSegment(g, s);
	    drawPosts(g);
	}

	void drawValuesOnLongestSegment(Graphics g, String s) {
	    // find longest segment
	    double bestLen = 0;
	    int bestSeg = 0;
	    for (int i = 0; i < routePoints.size() - 1; i++) {
		Point a = routePoints.get(i);
		Point b = routePoints.get(i + 1);
		double dx = b.x - a.x;
		double dy = b.y - a.y;
		double len = dx * dx + dy * dy;
		if (len > bestLen) {
		    bestLen = len;
		    bestSeg = i;
		}
	    }
	    Point a = routePoints.get(bestSeg);
	    Point b = routePoints.get(bestSeg + 1);
	    int mx = (a.x + b.x) / 2;
	    int my = (a.y + b.y) / 2;
	    g.setFont(unitsFont);
	    int w = (int)g.context.measureText(s).getWidth();
	    int ya = (int)g.currentFontSize/2;
	    g.setColor(whiteColor);
	    if (a.y == b.y)
		// horizontal segment: draw above
		g.drawString(s, mx - w/2, my - 6);
	    else
		// vertical segment: draw to the right
		g.drawString(s, mx + 4, my + ya);
	}

	void doDots(Graphics g) {
	    updateDotCount();
	    if (isCreating() || curcount == 0)
		return;
	    double cc = curcount;
	    for (int i = 0; i < routePoints.size() - 1; i++) {
		Point a = routePoints.get(i);
		Point b = routePoints.get(i + 1);
		drawDots(g, a, b, cc);
		// offset by segment length for continuity
		double dx = b.x - a.x;
		double dy = b.y - a.y;
		double segLen = Math.sqrt(dx * dx + dy * dy);
		cc = addCurCount(cc, segLen);
	    }
	}

	int getMouseDistance(int gx, int gy) {
	    int thresh = 10;
	    int best = Integer.MAX_VALUE;
	    for (int i = 0; i < routePoints.size() - 1; i++) {
		Point a = routePoints.get(i);
		Point b = routePoints.get(i + 1);
		int d = lineDistanceSq(a.x, a.y, b.x, b.y, gx, gy);
		if (d < best)
		    best = d;
	    }
	    if (best <= thresh * thresh)
		return best;
	    return -1;
	}

	// return the snapped point on the wire nearest to (mx, my), or null
	Point getSnapPointOnWire(int mx, int my) {
	    if (routePoints == null || routePoints.size() < 2)
		return null;

	    int bestSeg = -1;
	    int bestDist = Integer.MAX_VALUE;
	    for (int i = 0; i < routePoints.size() - 1; i++) {
		Point a = routePoints.get(i);
		Point b = routePoints.get(i + 1);
		int d = lineDistanceSq(a.x, a.y, b.x, b.y, mx, my);
		if (d < bestDist) {
		    bestDist = d;
		    bestSeg = i;
		}
	    }

	    Point a = routePoints.get(bestSeg);
	    Point b = routePoints.get(bestSeg + 1);

	    int sx, sy;
	    if (a.x == b.x) {
		sx = a.x;
		sy = snapGrid(my);
		//sy = Math.max(Math.min(a.y, b.y), Math.min(sy, Math.max(a.y, b.y)));
	    } else {
		sy = a.y;
		sx = snapGrid(mx);
		//sx = Math.max(Math.min(a.x, b.x), Math.min(sx, Math.max(a.x, b.x)));
	    }

	    return new Point(sx, sy);
	}

	void getInfo(String arr[]) {
	    super.getInfo(arr);
	    arr[0] = (busWidth > 1) ? "routed bus wire (" + busWidth + ")" : "routed wire";
	}
    }

File: RoutedWireElm.ts
/*
    Copyright (C) Paul Falstad and Iain Sharp

    This file is part of CircuitJS1.

    CircuitJS1 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    CircuitJS1 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CircuitJS1.  If not, see <http://www.gnu.org/licenses/>.
*/

import { WireElm } from "./WireElm";
import { WireRouter } from "./WireRouter";
import { CircuitElm } from "./CircuitElm";
import { Graphics } from "./Graphics";
import { Point } from "./Point";
import { StringTokenizer } from "./StringTokenizer";
import { CircuitXMLDeserializer } from "./CircuitXMLDeserializer";
import { CirSim } from "./CirSim";
import { UIManager } from "./UIManager";

export class RoutedWireElm extends WireElm {
    routePoints: Point[] | null = null;

    constructor(xx: number, yy: number);
    constructor(xa: number, ya: number, xb: number, yb: number, f: number, st: StringTokenizer);
    constructor(points: Point[]);
    constructor(xxOrPoints: number | Point[], yy?: number, xb?: number, yb?: number, f?: number, st?: StringTokenizer) {
        if (Array.isArray(xxOrPoints)) {
            const points = xxOrPoints;
            super(points[0].x, points[0].y);
            this.setPoints(points);
        } else if (xb !== undefined) {
            super(xxOrPoints, yy!, xb, yb!, f!, st!);
        } else {
            super(xxOrPoints, yy!);
        }
    }

    getDumpType(): number { return 0; }
    getXmlDumpType(): string { return "rw"; }

    dumpXml(doc: Document, elem: Element): void {
        super.dumpXml(doc, elem);
        if (this.routePoints !== null && this.routePoints.length > 0) {
            const parts = this.routePoints.map(p => p.x + "," + p.y);
            elem.appendChild(doc.createTextNode(parts.join(";")));
        }
    }

    undumpXml(xml: CircuitXMLDeserializer): void {
        super.undumpXml(xml);
        try {
            const contents = xml.parseContents();
            if (contents !== null && contents.length > 0) {
                const points: Point[] = [];
                for (const pair of contents.split(";")) {
                    const xy = pair.split(",");
                    points.push(new Point(parseInt(xy[0]), parseInt(xy[1])));
                }
                if (points.length >= 2)
                    this.routePoints = points;
            }
        } catch (e) {}
    }

    getShortcut(): number { return 'W'.charCodeAt(0); }

    setPoints(pointsOrRouting?: boolean | Point[]): void {
        if (Array.isArray(pointsOrRouting)) {
            const points = pointsOrRouting;
            const first = points[0];
            const last = points[points.length - 1];
            this.x = first.x; this.y = first.y;
            this.x2 = last.x; this.y2 = last.y;
            super.setPoints();
            this.routePoints = points;
            return;
        }

        const routing = pointsOrRouting !== false;
        super.setPoints();

        // if endpoints haven't changed, keep existing route
        if (this.routePoints !== null && this.routePoints.length >= 2) {
            const first = this.routePoints[0];
            const last = this.routePoints[this.routePoints.length - 1];
            if (first.x === this.x && first.y === this.y && last.x === this.x2 && last.y === this.y2)
                return;
        }

        if (!routing) {
            this.routePoints = [this.point1, this.point2];
            return;
        }

        const router = new WireRouter();
        router.initGrid(this, CirSim.theApp.gridSize, UIManager.theUI.elmList, UIManager.theUI.getCircuitBounds());
        const result = router.routeWire(this.x, this.y, this.x2, this.y2);
        if (result.length < 2) {
            CirSim.console("routing failed");
            this.routePoints = [this.point1, new Point(this.x2, this.y), this.point2];
            return;
        }
        CirSim.console("route success");
        this.routePoints = result;
    }

    // split this routed wire at the point nearest (mx, my)
    split(mx: number, my: number): RoutedWireElm | null {
        if (this.routePoints === null || this.routePoints.length < 2)
            return null;

        let bestSeg = -1;
        let bestDist = Number.MAX_SAFE_INTEGER;
        for (let i = 0; i < this.routePoints.length - 1; i++) {
            const a = this.routePoints[i];
            const b = this.routePoints[i + 1];
            const d = RoutedWireElm.segmentDistanceSq(a.x, a.y, b.x, b.y, mx, my);
            if (d < bestDist) { bestDist = d; bestSeg = i; }
        }

        const a = this.routePoints[bestSeg];
        const b = this.routePoints[bestSeg + 1];

        let sx: number, sy: number;
        if (a.x === b.x) {
            sx = a.x;
            sy = this.snapGrid(my);
            sy = Math.max(Math.min(a.y, b.y), Math.min(sy, Math.max(a.y, b.y)));
        } else {
            sy = a.y;
            sx = this.snapGrid(mx);
            sx = Math.max(Math.min(a.x, b.x), Math.min(sx, Math.max(a.x, b.x)));
        }

        if ((sx === this.x && sy === this.y) || (sx === this.x2 && sy === this.y2))
            return null;

        const rp1: Point[] = [];
        for (let i = 0; i <= bestSeg; i++) rp1.push(this.routePoints[i]);
        rp1.push(new Point(sx, sy));

        const rp2: Point[] = [new Point(sx, sy)];
        for (let i = bestSeg + 1; i < this.routePoints.length; i++) rp2.push(this.routePoints[i]);

        this.setPoints(rp1);
        return new RoutedWireElm(rp2);
    }

    rerouteVia(vx: number, vy: number): void {
        const router = new WireRouter();
        router.initGrid(this, CirSim.theApp.gridSize, UIManager.theUI.elmList, UIManager.theUI.getCircuitBounds());
        const rp1 = router.routeWire(this.x, this.y, vx, vy);
        if (rp1.length < 2) return;

        for (let i = 0; i < rp1.length - 1; i++)
            router.addWire(rp1[i].x, rp1[i].y, rp1[i + 1].x, rp1[i + 1].y);

        const rp2 = router.routeWire(vx, vy, this.x2, this.y2);
        if (rp2.length < 2) return;

        this.routePoints = [...rp1, ...rp2.slice(1)];
    }

    addRoutingObstacle(router: WireRouter): void {
        if (this.routePoints === null) return;
        for (let i = 0; i < this.routePoints.length - 1; i++)
            router.addWire(this.routePoints[i].x, this.routePoints[i].y,
                           this.routePoints[i + 1].x, this.routePoints[i + 1].y);
    }

    draw(g: Graphics): void {
        if (this.routePoints === null || this.routePoints.length < 2) {
            super.draw(g);
            return;
        }
        this.setVoltageColor(g, this.volts[0]);
        for (let i = 0; i < this.routePoints.length - 1; i++)
            CircuitElm.drawThickLine(g, this.routePoints[i], this.routePoints[i + 1], this.busWidth > 1 ? 5 : 3);

        if (this.currents !== null) {
            this.current = 0;
            for (let i = 0; i < this.currents.length; i++)
                this.current += this.currents[i];
        }
        this.doDots(g);

        const m = 5;
        let minX = this.point1.x, minY = this.point1.y, maxX = this.point1.x, maxY = this.point1.y;
        for (let i = 1; i < this.routePoints.length; i++) {
            const p = this.routePoints[i];
            minX = Math.min(minX, p.x); minY = Math.min(minY, p.y);
            maxX = Math.max(maxX, p.x); maxY = Math.max(maxY, p.y);
        }
        this.setBbox(minX - m, minY - m, maxX + m, maxY + m);

        let s = "";
        if (this.busWidth > 1 && (this.mustShowBusValue() || this.mustShowBusValueHex())) {
            const value = this.getBusValue();
            if (this.mustShowBusValue()) s = "" + value;
            if (this.mustShowBusValueHex()) s = (s.length > 0 ? s + " " : "") + "0x" + value.toString(16).toUpperCase();
        } else if (this.busWidth === 1) {
            if (this.mustShowCurrent()) s = CircuitElm.getShortUnitText(Math.abs(this.getCurrent()), "A");
            if (this.mustShowVoltage()) s = (s.length > 0 ? s + " " : "") + CircuitElm.getShortUnitText(this.volts[0], "V");
        }
        if (s.length > 0)
            this.drawValuesOnLongestSegment(g, s);
        this.drawPosts(g);
    }

    drawValuesOnLongestSegment(g: Graphics, s: string): void {
        if (this.routePoints === null || this.routePoints.length < 2) return;
        let bestLen = 0, bestSeg = 0;
        for (let i = 0; i < this.routePoints.length - 1; i++) {
            const a = this.routePoints[i], b = this.routePoints[i + 1];
            const dx = b.x - a.x, dy = b.y - a.y;
            const len = dx * dx + dy * dy;
            if (len > bestLen) { bestLen = len; bestSeg = i; }
        }
        const a = this.routePoints[bestSeg], b = this.routePoints[bestSeg + 1];
        const mx = (a.x + b.x) / 2 | 0, my = (a.y + b.y) / 2 | 0;
        g.setFont(CircuitElm.unitsFont);
        const w = g.measureWidth(s);
        const ya = g.currentFontSize / 2 | 0;
        g.setColor(CircuitElm.whiteColor);
        if (a.y === b.y)
            g.drawString(s, mx - w / 2, my - 6);
        else
            g.drawString(s, mx + 4, my + ya);
    }

    doDots(g: Graphics): void {
        this.updateDotCount();
        if (this.isCreating() || this.curcount === 0 || this.routePoints === null) return;
        let cc = this.curcount;
        for (let i = 0; i < this.routePoints.length - 1; i++) {
            const a = this.routePoints[i], b = this.routePoints[i + 1];
            this.drawDots(g, a, b, cc);
            const dx = b.x - a.x, dy = b.y - a.y;
            const segLen = Math.sqrt(dx * dx + dy * dy);
            cc = this.addCurCount(cc, segLen);
        }
    }

    getMouseDistance(gx: number, gy: number): number {
        if (this.routePoints === null) return super.getMouseDistance(gx, gy);
        const thresh = 10;
        let best = Number.MAX_SAFE_INTEGER;
        for (let i = 0; i < this.routePoints.length - 1; i++) {
            const a = this.routePoints[i], b = this.routePoints[i + 1];
            const d = RoutedWireElm.segmentDistanceSq(a.x, a.y, b.x, b.y, gx, gy);
            if (d < best) best = d;
        }
        return best <= thresh * thresh ? best : -1;
    }

    static segmentDistanceSq(ax: number, ay: number, bx: number, by: number, gx: number, gy: number): number {
        const dx = bx - ax, dy = by - ay;
        const lenSq = dx * dx + dy * dy;
        if (lenSq === 0) return (gx - ax) * (gx - ax) + (gy - ay) * (gy - ay);
        let t = ((gx - ax) * dx + (gy - ay) * dy) / lenSq;
        if (t < 0) t = 0; else if (t > 1) t = 1;
        const px = ax + t * dx, py = ay + t * dy;
        const ex = gx - px, ey = gy - py;
        return ex * ex + ey * ey | 0;
    }

    pointOnPath(p: Point): boolean {
        if (this.routePoints === null || this.routePoints.length < 2) return false;
        if (p.equals(this.point1) || p.equals(this.point2)) return false;
        for (let i = 0; i < this.routePoints.length - 1; i++) {
            const a = this.routePoints[i], b = this.routePoints[i + 1];
            if (a.x === b.x && p.x === a.x) {
                const lo = Math.min(a.y, b.y), hi = Math.max(a.y, b.y);
                if (p.y >= lo && p.y <= hi) return true;
            } else if (a.y === b.y && p.y === a.y) {
                const lo = Math.min(a.x, b.x), hi = Math.max(a.x, b.x);
                if (p.x >= lo && p.x <= hi) return true;
            }
        }
        return false;
    }

    getSnapPointOnWire(mx: number, my: number): Point | null {
        if (this.routePoints === null || this.routePoints.length < 2) return null;

        let bestSeg = -1, bestDist = Number.MAX_SAFE_INTEGER;
        for (let i = 0; i < this.routePoints.length - 1; i++) {
            const a = this.routePoints[i], b = this.routePoints[i + 1];
            const d = this.lineDistanceSq(a.x, a.y, b.x, b.y, mx, my);
            if (d < bestDist) { bestDist = d; bestSeg = i; }
        }

        const a = this.routePoints[bestSeg], b = this.routePoints[bestSeg + 1];
        let sx: number, sy: number;
        if (a.x === b.x) {
            sx = a.x;
            sy = this.snapGrid(my);
        } else {
            sy = a.y;
            sx = this.snapGrid(mx);
        }
        return new Point(sx, sy);
    }

    getInfo(arr: string[]): void {
        super.getInfo(arr);
        arr[0] = (this.busWidth > 1) ? "routed bus wire (" + this.busWidth + ")" : "routed wire";
    }
}
